packages/playwright-test/src/testType.ts

```diff
@@ -202,6 +208,29 @@ export class TestTypeImpl {
     }
   }
 
+  private async _attach(location: Location, attachment: TestAttachment) {
+    const testInfo = currentTestInfo();
+    if (!testInfo)
+      throw errorWithLocation(location, `test.attach() can only be called from a test`);
+
+    if (attachment.body && attachment.path)
+      throw errorWithLocation(location, `test.attach() does not allow both body and path to be specificed, but both were. Use one or the other`);
+
+    if (attachment.body === undefined && attachment.path === undefined)
+      errorWithLocation(location, `test.attach() expects either body or path to be defined, but neither was`);
+
+    const tmpAttachment = { ...attachment };
+    if (attachment.path) {
+      const hash = await calculateFileSha1(attachment.path);
+      const dest = testInfo.outputPath('attachments', hash + path.extname(attachment.path));
+      await fs.promises.mkdir(path.dirname(dest), { recursive: true });
+      await fs.promises.copyFile(attachment.path, dest);
+      tmpAttachment.path = dest;
```

Comment:
NB: This is intentionally an absolute path. Reporters are responsible for "relativising" it as needed for their packaging formats.